### PR TITLE
Introduce `TypeUrl.ofTypeOrService()`

### PR DESCRIPTION
**Pull request by alexander-yevsyukov**
This PR introduces the factory method `TypeUrl.ofTypeOrService()`, revering the [introduction of `TypeUrl.of()`](#839). 

The problem with `TypeUrl.of()` was that accepting `GenericDescriptor` in the method with such a popular name was confusing and resulted in lots of runtime errors after adoption in ProtoData. We still need a method which accepts `GenericDescriptor` but in a method with more descriptive name.

The implementation of `TypeUrl.ofTypeOrService()` now explicitly checks for the type of the passed argument with clear error diagnostics.

## Other notable changes
 * The extension `Descriptor.realNestedTypes()` recently used in ProtoData was added to `DescriptorExts.kt`.
 * `TypeUrl.toTypeName()` was deprecated in favour of new method `typeName()`. We want to avoid the `to` prefix because the method simply returns the field without making any conversion.
 * Tests of `TypeUrl` were converted to Kotlin.
 * Fixed names of test proto files that used dashes instead of underscores in the file names.

### CODE REVIEW

File: src/test/proto/spine/test/protobuf/descriptor_exts_spec.proto [hunk 1]
```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright 2022, TeamDev. All rights reserved.
```
Comment (armiol) [inline src/test/proto/spine/test/protobuf/descriptor_exts_spec.proto:2]: 2024.

File: src/test/proto/spine/test/protobuf/descriptor_exts_spec.proto [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
```
Comment (armiol) [inline src/test/proto/spine/test/protobuf/descriptor_exts_spec.proto:8]: S.

**Review by armiol — APPROVED** [review]
@alexander-yevsyukov LGTM with the regards to two minor comments I left.

File: src/test/kotlin/io/spine/protobuf/DescriptorExtsSpec.kt [hunk 3]
```diff
@@ -55,4 +58,20 @@ internal class DescriptorExtsSpec {
             descriptor.field(SECONDS_FIELD_NUMBER)!!.number shouldBe SECONDS_FIELD_NUMBER
         }
     }
+
+    @Test
+    fun `list nested types excluding synthetic ones for map entries`() {
+        val descr = AddressBook.getDescriptor()
```
Comment (yevhenii-nadtochii) [inline src/test/kotlin/io/spine/protobuf/DescriptorExtsSpec.kt:64]: What if `AddressBook` has a nested `RecentEntry` message, occupying the name of the map entry?

File: src/test/kotlin/io/spine/type/TypeUrlSpec.kt [hunk 4]
```diff
@@ -0,0 +1,329 @@
...
+    private val typeName: String = descriptor.fullName
+
+    /** The value of the Type URL.  */
+    private val typeUrlValue: String =
+        TypeUrl.composeTypeUrl(TypeUrl.Prefix.GOOGLE_APIS.value(), typeName)
+
+    private val stringValueTypeUrl: TypeUrl = TypeUrl.from(descriptor)
+
+    @Nested
+    internal inner class
```
Comment (yevhenii-nadtochii) [inline src/test/kotlin/io/spine/type/TypeUrlSpec.kt:77]: It was re-formatted here and below.

File: src/test/kotlin/io/spine/protobuf/DescriptorExtsSpec.kt (hunk 3 above)
Comment (alexander-yevsyukov) [reply src/test/kotlin/io/spine/protobuf/DescriptorExtsSpec.kt:64]: This is prohibited by [Protobuf specs](https://protobuf.dev/programming-guides/proto3/#maps-features).

**Pull request merged**